packages/core/src/deal-tracker/spade-oracle-sync-tick.js

```diff
@@ -0,0 +1,238 @@
+  const res = await Promise.all(
+    Array.from(diffPieceContracts, ([key, value]) => {
+      return Promise.all(value.map(contract => {
+        /** @type {import('@web3-storage/data-segment').LegacyPieceLink} */
+        const legacyPieceCid = parseLink(key)
+
+        return dealStore.put({
+          ...contract,
+          // @ts-expect-error not PieceCIDv2
+          piece: legacyPieceCid,
+          provider: `${contract.provider}`,
+          insertedAt: (new Date()).toISOString(),
+          updatedAt: (new Date()).toISOString()
+        })
+      }))
+    })
+  )
+
+  const firsPutError = res.find(pieceContracts => pieceContracts.find(c => c.error))?.find(comb => comb.error)
+  if (firsPutError?.error) {
+    return {
+      error: firsPutError.error
+    }
+  }
+  return {
+    ok: {}
+  }
+}
+
+/**
+ * @param {object} context
+ * @param {PieceContracts} context.currentPieceContracts 
+ * @param {PieceContracts} context.updatedPieceContracts 
+ */
+export function computeDiff ({ currentPieceContracts, updatedPieceContracts }) {
+  /** @type {PieceContracts} */
+  const diff = new Map()
+
+  for (const [pieceCid, contracts] of updatedPieceContracts.entries() ) {
+    const currentContracts = currentPieceContracts.get(pieceCid) || []
+    // Find diff when different length
+    if (contracts.length !== currentContracts.length) {
+      const diffContracts = []
+      // Get contracts for PieceCID still not recorded
+      for (const c of contracts) {
+        if (!currentContracts.find(pc => pc.dealId === c.dealId)) {
+          diffContracts.push(c)
+        }
+      }
+      diff.set(pieceCid, diffContracts)
+    }
+  }
+
+  return diff
+}
+
+/**
+ * @param {object} context
+ * @param {DealArchiveStore} context.dealArchiveStore
+ * @param {string} context.spadeOracleId
+ * @returns {Promise<import('../types').Result<PieceContracts, Error>>}
+ */
+export async function getCurrentDealArchive ({ dealArchiveStore, spadeOracleId }) {
+  const getRes = await dealArchiveStore.get(spadeOracleId)
+  if (getRes.error) {
+    return getRes
+  }
+
+  return {
+    ok: new Map(Object.entries(decode(getRes.ok.value))),
+  }
+}
+
+/**
+ * @param {object} context
+ * @param {DealArchiveStore} context.dealArchiveStore
+ * @param {string} context.spadeOracleId
+ * @param {PieceContracts} context.oracleContracts 
+ */
+async function putLatestDealArchive ({ dealArchiveStore, spadeOracleId, oracleContracts }) {
+  const putRes = await dealArchiveStore.put({
+    key: spadeOracleId,
+    value: encode(Object.fromEntries(oracleContracts))
+  })
+
+  return putRes
+}
+
+/**
+ * @param {URL} spadeOracleUrl
+ * @returns {Promise<import('../types').Result<PieceContracts, Error>>}
+ */
+async function fetchLatestDealArchive (spadeOracleUrl) {
+  /** @type {PieceContracts} */
+  const dealMap = new Map()
+  const res = await fetch(spadeOracleUrl)
+  if (!res.ok) {
+    return {
+      // TODO: Error
+      error: new Error('could not read')
```

Comment:
I typically just report the status code:
```suggestion
      error: new Error(`unexpected response status fetching deal archive: ${res.status}`)
```